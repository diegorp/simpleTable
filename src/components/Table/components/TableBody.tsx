import type { Column } from '@/types';

interface Props<T> {
  rows: T[];
  columns: Column<T>[];
  rowIdKey: keyof T;
}

export default function TableBody<T>({ rows, columns, rowIdKey }: Props<T>) {
  return (
    <tbody className='bg-white'>
      {rows.length === 0 ? (
        <tr className='border-b last:border-b-0 border-[#E2E8F0]'>
          <td colSpan={columns.length} className='px-6 py-4 text-center'>
            No data to display.
          </td>
        </tr>
      ) : (
        rows.map(row => (
          <tr
            key={row[rowIdKey] as string}
            className='border-b last:border-b-0 border-[#E2E8F0] hover:bg-gray-50 transition-colors duration-200 ease-in-out'
          >
            {columns.map(column => (
              <td
                key={column.key as string}
                className='px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]'
              >
                {column.render(row[column.key])}
              </td>
            ))}
          </tr>
        ))
      )}
    </tbody>
  );
}
