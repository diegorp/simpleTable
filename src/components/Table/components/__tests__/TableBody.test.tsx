import { render } from '@testing-library/react';
import TableBody from '@/components/Table/components/TableBody';
import type { Column } from '@/types';

describe('<TableBody />', () => {
  const mockColumns: Column<{ id: number; name: string; age: number }>[] = [
    { key: 'id', label: 'ID', render: (value) => String(value) },
    { key: 'name', label: 'Name', render: (value) => String(value) },
    { key: 'age', label: 'Age', render: (value) => String(value) },
  ];

  const mockRows = [
    { id: 1, name: 'Alice', age: 25 },
    { id: 2, name: 'Bob', age: 30 },
  ];

  it('should match snapshot', () => {
    const { container } = render(
      <table>
        <TableBody rows={mockRows} columns={mockColumns} rowIdKey="id" />
      </table>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render all rows', () => {
    const { getByText } = render(
      <table>
        <TableBody rows={mockRows} columns={mockColumns} rowIdKey="id" />
      </table>
    );

    expect(getByText('Alice')).toBeInTheDocument();
    expect(getByText('Bob')).toBeInTheDocument();
  });

  it('should render all columns for each row', () => {
    const { getByText } = render(
      <table>
        <TableBody rows={mockRows} columns={mockColumns} rowIdKey="id" />
      </table>
    );

    expect(getByText('1')).toBeInTheDocument();
    expect(getByText('Alice')).toBeInTheDocument();
    expect(getByText('25')).toBeInTheDocument();
    expect(getByText('2')).toBeInTheDocument();
    expect(getByText('Bob')).toBeInTheDocument();
    expect(getByText('30')).toBeInTheDocument();
  });

  it('should display "No data to display" when rows are empty', () => {
    const { getByText } = render(
      <table>
        <TableBody rows={[]} columns={mockColumns} rowIdKey="id" />
      </table>
    );

    expect(getByText('No data to display.')).toBeInTheDocument();
  });

  it('should use custom render function for columns', () => {
    const customColumns: Column<{ id: number; name: string; age: number }>[] = [
      { key: 'id', label: 'ID', render: (value) => `ID: ${value}` },
      { key: 'name', label: 'Name', render: (value) => String(value).toUpperCase() },
      { key: 'age', label: 'Age', render: (value) => `${value} years` },
    ];

    const { getByText } = render(
      <table>
        <TableBody rows={mockRows} columns={customColumns} rowIdKey="id" />
      </table>
    );

    expect(getByText('ID: 1')).toBeInTheDocument();
    expect(getByText('ALICE')).toBeInTheDocument();
    expect(getByText('25 years')).toBeInTheDocument();
  });

  it('should render correct number of rows', () => {
    const { container } = render(
      <table>
        <TableBody rows={mockRows} columns={mockColumns} rowIdKey="id" />
      </table>
    );

    const rows = container.querySelectorAll('tbody > tr');
    expect(rows).toHaveLength(2);
  });

  it('should render correct number of cells per row', () => {
    const { container } = render(
      <table>
        <TableBody rows={mockRows} columns={mockColumns} rowIdKey="id" />
      </table>
    );

    const firstRow = container.querySelector('tbody > tr:first-child');
    const cells = firstRow?.querySelectorAll('td');
    expect(cells).toHaveLength(3);
  });

  it('should handle single row', () => {
    const { getByText, queryByText } = render(
      <table>
        <TableBody
          rows={[mockRows[0]]}
          columns={mockColumns}
          rowIdKey="id"
        />
      </table>
    );

    expect(getByText('Alice')).toBeInTheDocument();
    expect(queryByText('Bob')).not.toBeInTheDocument();
  });

  it('should span all columns in empty state', () => {
    const { container } = render(
      <table>
        <TableBody rows={[]} columns={mockColumns} rowIdKey="id" />
      </table>
    );

    const emptyCell = container.querySelector('td[colspan="3"]');
    expect(emptyCell).toBeInTheDocument();
  });
});
