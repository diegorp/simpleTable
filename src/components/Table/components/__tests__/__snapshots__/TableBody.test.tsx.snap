// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<TableBody /> > should match snapshot 1`] = `
<div>
  <table>
    <tbody
      class="bg-white"
    >
      <tr
        class="border-b last:border-b-0 border-[#E2E8F0] hover:bg-gray-50 transition-colors duration-200 ease-in-out"
      >
        <td
          class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
        >
          1
        </td>
        <td
          class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
        >
          Alice
        </td>
        <td
          class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
        >
          25
        </td>
      </tr>
      <tr
        class="border-b last:border-b-0 border-[#E2E8F0] hover:bg-gray-50 transition-colors duration-200 ease-in-out"
      >
        <td
          class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
        >
          2
        </td>
        <td
          class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
        >
          Bob
        </td>
        <td
          class="px-6 py-4 whitespace-nowrap border-r last:border-r-0 border-[#E2E8F0]"
        >
          30
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
